clang/lib/AST/ItaniumMangle.cpp

```diff
@@ -2503,7 +2543,20 @@ void CXXNameMangler::mangleType(QualType T) {
   const Type *ty = split.Ty;
 
   bool isSubstitutable =
-    isTypeSubstitutable(quals, ty, Context.getASTContext());
+      isTypeSubstitutable(quals, ty, Context.getASTContext());
+  if (Context.isUniqueNameMangler()) {
+    static std::array<DeclContextDesc, 5> Scopes = {
+        DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
+        DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
+        DeclContextDesc{clang::Decl::Kind::Namespace, "detail"},
+        DeclContextDesc{clang::Decl::Kind::Namespace, "half_impl"},
```

Comment:
What if we change names of classes/namespaces in SYCL headers?
This seems even worse than the way which we use to detect accessors, because name and namespaces of accessors is defined by SYCL spec. This set of detail/half_impl anything is NOT defined by the SYCL spec, it's only details of our implementation of SYCL headers.
This "magic" is fragile, not flexible and cannot be upstreamed.

Reply:
> What if we change names of classes/namespaces in SYCL headers?

Corresponding LIT test will fail and we will have to update this code too.

> This "magic" is fragile, not flexible and cannot be upstreamed.

I agree that the whole solution looks very "hacky". Any ideas how to do the same better?

Reply:
We could try to define `cl::sycl::half` as a single wrapper. But change underlying type depending on device we or not. Then it will be the same type for mangler. Same as it's done for `cl::sycl::vec` classes.

Reply:
What do you mean 'as a single wrapper'?  As a type itself?  Type aliases aren't mangled.  Curious to see what the 'vec' classes solution is.

Reply:
@erichkeane, the idea is to use `cl::sycl::detail::half_impl::half` for both device and host code to achieve the same mangling.

This might require writing relatively significant amount of code somewhere in SYCL headers, because this wrapper should encapsulate native half within it on device side:
- device compiler should be able to optimize math operations with it
- we should be able to call different device built-in functions which accept native half

I will try to do this, but I also would like to get "LGTM" for this patch as back-up plan if the different way will take a lot of time